File: src/services/categorization/ProductParsingService.ts
import {ParsingResult} from "../../types/product";
import {UnitService} from "./UnitService";
import {QuantityParser} from "./QuantityParser";

export class ProductParsingService {

    /*
        private static readonly KITCHEN_MEASURES: Record<string, number> = {
            'łyżka': 15,
            'łyżeczka': 5,
            'szklanka': 250,
            'garść': 30
        };
    */

    static parseProduct(input: string): ParsingResult {
        try {
            const cleanInput = this.cleanInputString(input);
            const quantityInfo = this.extractQuantityAndUnit(cleanInput);

            if (!quantityInfo.success || !quantityInfo.quantity || !quantityInfo.remainingText) {
                return {
                    success: false,
                    error: 'Nie znaleziono poprawnej ilości'
                };
            }

            const {quantity, potentialUnit, remainingText} = quantityInfo;

            if (!potentialUnit && !remainingText) {
                const nameParts = cleanInput.split(' ').filter(part => !this.isNumeric(part));
                return {
                    success: true,
                    product: {
                        name: this.cleanProductName(nameParts.join(' ')),
                        quantity,
                        unit: 'szt',
                        original: input,
                        hasCustomUnit: false
                    }
                };
            }

            const {unit, name, foundKnownUnit} = this.processUnitAndName(potentialUnit || '', remainingText);

            // Normalizuj wartość do jednostki bazowej
            const normalized = UnitService.normalizeToBaseUnit(quantity, unit);

            if (normalized) {
                return {
                    success: true,
                    product: {
                        name: this.cleanProductName(name),
                        quantity: normalized.value,
                        unit: normalized.unit,
                        original: input,
                        hasCustomUnit: !foundKnownUnit && !UnitService.isValidUnit(unit)
                    }
                };
            }

            // Jeśli nie można znormalizować, zwróć oryginalne wartości
            return {
                success: true,
                product: {
                    name: this.cleanProductName(name),
                    quantity,
                    unit,
                    original: input,
                    hasCustomUnit: !foundKnownUnit && !UnitService.isValidUnit(unit)
                }
            };

        } catch (error) {
            return {
                success: false,
                error: error instanceof Error ? error.message : 'Nieznany błąd parsowania'
            };
        }
    }

    private static cleanInputString(input: string): string {
        return input
            .trim()
            .replace(/^[•\-]\s*/, '')  // Usuń znaki wypunktowania
            .replace(/\s+/g, ' ');     // Normalizuj spacje
    }

    private static extractQuantityAndUnit(input: string): {
        success: boolean;
        quantity?: number;
        potentialUnit?: string;
        remainingText?: string;
    } {
        const patterns = [
            // Pattern dla "pół kg mąki", "2-3 kg mąki", "1/2 kg mąki"
            /^(pół|półtorej|ćwierć|jedna|jeden|dwa|trzy|cztery|pięć|\d+(?:[.,]\d+)?(?:\s*-\s*\d+(?:[.,]\d+)?)?|\d+\/\d+|\d+\s+\d+\/\d+)\s*([^\d\s]+)?\s*(.+)?/i,
            // Pattern dla "mąka pół kg", "mąka 2-3 kg"
            /^(.+?)\s+(pół|półtorej|ćwierć|jedna|jeden|dwa|trzy|cztery|pięć|\d+(?:[.,]\d+)?(?:\s*-\s*\d+(?:[.,]\d+)?)?|\d+\/\d+|\d+\s+\d+\/\d+)\s*([^\d\s]+)?$/i,
        ];

        for (const pattern of patterns) {
            const match = input.match(pattern);
            if (match) {
                let quantityStr: string;
                let potentialUnit: string | undefined;
                let remainingText: string | undefined;

                if (pattern === patterns[0]) {
                    // Format "2-3 kg mąki"
                    quantityStr = match[1];
                    potentialUnit = match[2]?.toLowerCase();
                    remainingText = match[3];
                } else {
                    // Format "mąka 2-3 kg"
                    quantityStr = match[2];
                    potentialUnit = match[3]?.toLowerCase();
                    remainingText = match[1];
                }

                const quantity = QuantityParser.parseQuantity(quantityStr);
                if (quantity === null) continue;

                if (!remainingText && potentialUnit) {
                    remainingText = potentialUnit;
                    potentialUnit = undefined;
                }

                return {
                    success: true,
                    quantity,
                    potentialUnit,
                    remainingText: remainingText?.trim()
                };
            }
        }

        return {success: false};
    }

    private static processUnitAndName(potentialUnit: string | undefined, remainingText: string): {
        unit: string;
        name: string;
        foundKnownUnit: boolean;
    } {
        let unit = '';
        let name = remainingText;
        let foundKnownUnit = false;

        // Najpierw sprawdź potencjalną jednostkę
        if (potentialUnit) {
            const unitInfo = UnitService.detectUnitInText(potentialUnit);
            if (unitInfo.match) {
                unit = unitInfo.unit;
                foundKnownUnit = true;
            }
        }

        // Jeśli nie znaleziono jednostki w potencjalnej jednostce,
        // sprawdź w pozostałym tekście
        if (!foundKnownUnit && remainingText) {
            const unitInfo = UnitService.detectUnitInText(remainingText);
            if (unitInfo.match) {
                unit = unitInfo.unit;
                foundKnownUnit = true;
                // Usuń jednostkę z nazwy produktu
                name = remainingText.replace(new RegExp(`\\b${unit}\\b`, 'i'), '').trim();
            }
        }

        // Ustaw jednostkę domyślną, tylko jeśli naprawdę nie znaleziono żadnej innej
        if (!unit || unit === '') {
            unit = 'szt';
            foundKnownUnit = true;
        }

        return {unit, name, foundKnownUnit};
    }

    static cleanProductName(name: string): string {
        return name
            .replace(/\s*\([^)]+\)/g, '')  // Usuń nawiasy z zawartością
            .replace(/\s+/g, ' ')           // Normalizuj spacje
            .trim()
            .toLowerCase();                 // Konwertuj na małe litery dla spójności
    }

    static calculateSimilarity(product1: string, product2: string): number {
        const name1 = this.cleanProductName(product1);
        const name2 = this.cleanProductName(product2);

        // Użyj algorytmu Levenshtein distance do porównania nazw
        const distance = this.calculateLevenshteinDistance(name1, name2);
        return 1 - (distance / Math.max(name1.length, name2.length));
    }

    private static calculateLevenshteinDistance(str1: string, str2: string): number {
        const matrix: number[][] = [];

        for (let i = 0; i <= str1.length; i++) {
            matrix[i] = [i];
        }

        for (let j = 0; j <= str2.length; j++) {
            matrix[0][j] = j;
        }

        for (let i = 1; i <= str1.length; i++) {
            for (let j = 1; j <= str2.length; j++) {
                if (str1[i - 1] === str2[j - 1]) {
                    matrix[i][j] = matrix[i - 1][j - 1];
                } else {
                    matrix[i][j] = Math.min(
                        matrix[i - 1][j - 1] + 1,
                        matrix[i][j - 1] + 1,
                        matrix[i - 1][j] + 1
                    );
                }
            }
        }

        return matrix[str1.length][str2.length];
    }

    private static isNumeric(str: string): boolean {
        return !isNaN(parseFloat(str)) && isFinite(Number(str));
    }
}

File: backend/src/main/java/com/noisevisionsoftware/nutrilog/utils/excelParser/service/ProductParsingService.java
package com.noisevisionsoftware.nutrilog.utils.excelParser.service;

import com.noisevisionsoftware.nutrilog.utils.excelParser.model.ParsedProduct;
import com.noisevisionsoftware.nutrilog.utils.excelParser.model.ParsingResult;
import com.noisevisionsoftware.nutrilog.utils.excelParser.model.QuantityInfo;
import com.noisevisionsoftware.nutrilog.utils.excelParser.model.UnitProcessingResult;
import com.noisevisionsoftware.nutrilog.utils.excelParser.model.unit.UnitDetectionResult;
import com.noisevisionsoftware.nutrilog.utils.excelParser.service.helpers.QuantityParser;
import com.noisevisionsoftware.nutrilog.utils.excelParser.service.helpers.UnitService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProductParsingService {

    private final UnitService unitService;
    private final QuantityParser quantityParser;

    // Wzorce dla różnych formatów ilości
    private static final List<Pattern> QUANTITY_PATTERNS = Arrays.asList(
            // Podstawowe wzorce, które już masz
            Pattern.compile("^(pół|półtorej|ćwierć|jedna|jeden|dwa|trzy|cztery|pięć|\\d+(?:[.,]\\d+)?(?:\\s*-\\s*\\d+(?:[.,]\\d+)?)?|\\d+/\\d+|\\d+\\s+\\d+/\\d+)\\s*([^\\d\\s]+)?\\s*(.+)?", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^(.+?)\\s+(pół|półtorej|ćwierć|jedna|jeden|dwa|trzy|cztery|pięć|\\d+(?:[.,]\\d+)?(?:\\s*-\\s*\\d+(?:[.,]\\d+)?)?|\\d+/\\d+|\\d+\\s+\\d+/\\d+)\\s*([^\\d\\s]+)?$", Pattern.CASE_INSENSITIVE),

            // Dodatkowe wzorce dla większej elastyczności
            Pattern.compile("^(\\d+)\\s*([^\\d\\s]+)\\s+(.+)$", Pattern.CASE_INSENSITIVE),  // "600 g marchwi"
            Pattern.compile("^(.+?)\\s+(\\d+)\\s*([^\\d\\s]+)$", Pattern.CASE_INSENSITIVE)  // "marchew 600 g"
    );

    public ParsingResult parseProduct(String input) {
        try {
            String cleanInput = cleanInputString(input);

            Pattern directPattern1 = Pattern.compile("^(\\d+)\\s*([a-ząćęłńóśźż]+)\\s+(.+)$", Pattern.CASE_INSENSITIVE);
            Matcher directMatcher1 = directPattern1.matcher(cleanInput);
            if (directMatcher1.matches()) {
                double quantity = Double.parseDouble(directMatcher1.group(1));
                String unit = directMatcher1.group(2).toLowerCase();
                String name = directMatcher1.group(3);

                return new ParsingResult(new ParsedProduct(
                        cleanProductName(name),
                        quantity,
                        unit,
                        input,
                        false
                ));
            }

            Pattern directPattern2 = Pattern.compile("^(.+?)\\s+(\\d+)\\s*([a-ząćęłńóśźż]+)$", Pattern.CASE_INSENSITIVE);
            Matcher directMatcher2 = directPattern2.matcher(cleanInput);
            if (directMatcher2.matches()) {
                String name = directMatcher2.group(1);
                double quantity = Double.parseDouble(directMatcher2.group(2));
                String unit = directMatcher2.group(3).toLowerCase();

                return new ParsingResult(new ParsedProduct(
                        cleanProductName(name),
                        quantity,
                        unit,
                        input,
                        false
                ));
            }

            QuantityInfo quantityInfo = extractQuantityAndUnit(cleanInput);

            if (!quantityInfo.isSuccess() || quantityInfo.getQuantity() == null || quantityInfo.getRemainingText() == null) {

                Pattern numberPattern = Pattern.compile("(\\d+(?:[.,]\\d+)?)");
                Matcher numberMatcher = numberPattern.matcher(cleanInput);
                if (numberMatcher.find()) {
                    double quantity = Double.parseDouble(numberMatcher.group(1).replace(',', '.'));
                    return new ParsingResult(new ParsedProduct(
                            cleanProductName(cleanInput.replace(numberMatcher.group(1), "").trim()),
                            quantity,
                            "szt",
                            input,
                            false
                    ));
                }

                return new ParsingResult(new ParsedProduct(
                        cleanProductName(cleanInput),
                        1.0,
                        "szt",
                        input,
                        false
                ));
            }

            double quantity = quantityInfo.getQuantity();
            String potentialUnit = quantityInfo.getPotentialUnit();
            String remainingText = quantityInfo.getRemainingText();

            if ((potentialUnit == null || potentialUnit.isEmpty()) && remainingText.isEmpty()) {
                String[] nameParts = cleanInput.split(" ");
                StringBuilder nameBuilder = new StringBuilder();
                for (String part : nameParts) {
                    if (!isNumeric(part)) {
                        nameBuilder.append(part).append(" ");
                    }
                }

                return new ParsingResult(new ParsedProduct(
                        cleanProductName(nameBuilder.toString().trim()),
                        quantity,
                        "szt",
                        input,
                        false
                ));
            }

            UnitProcessingResult processedUnit = processUnitAndName(
                    potentialUnit,
                    remainingText
            );

            // Jeśli nie można znormalizować, zwróć oryginalne wartości
            return new ParsingResult(new ParsedProduct(
                    cleanProductName(processedUnit.getName()),
                    quantity,
                    processedUnit.getUnit(),
                    input,
                    !processedUnit.isFoundKnownUnit() && unitService.isValidUnit(processedUnit.getUnit())
            ));
        } catch (Exception e) {
            return new ParsingResult(new ParsedProduct(
                    cleanInputString(input),
                    1.0,
                    "szt",
                    input,
                    false
            ));
        }
    }

    private String cleanInputString(String input) {
        return input
                .trim()
                .replaceAll("^[•\\-]\\s*", "")  // Usuń znaki wypunktowania
                .replaceAll("\\s+", " ");       // Normalizuj spacje
    }

    private QuantityInfo extractQuantityAndUnit(String input) {
        for (Pattern pattern : QUANTITY_PATTERNS) {
            Matcher matcher = pattern.matcher(input);
            if (matcher.find()) {
                String quantityStr;
                String potentialUnit;
                String remainingText;

                if (pattern == QUANTITY_PATTERNS.getFirst()) {
                    // Format "2-3 kg mąki"
                    quantityStr = matcher.group(1);
                    potentialUnit = matcher.group(2) != null ? matcher.group(2).toLowerCase() : null;
                    remainingText = matcher.group(3);
                } else {
                    // Format "mąka 2-3 kg"
                    quantityStr = matcher.group(2);
                    potentialUnit = matcher.group(3) != null ? matcher.group(3).toLowerCase() : null;
                    remainingText = matcher.group(1);
                }

                Double quantity = quantityParser.parseQuantity(quantityStr);
                if (quantity == null) continue;

                if (remainingText == null && potentialUnit != null) {
                    remainingText = potentialUnit;
                    potentialUnit = null;
                }

                if (remainingText != null) {
                    return new QuantityInfo(quantity, potentialUnit, remainingText.trim());
                }
            }
        }

        return new QuantityInfo(false);
    }

    private UnitProcessingResult processUnitAndName(String potentialUnit, String remainingText) {
        String unit = "";
        String name = remainingText;
        boolean foundKnownUnit = false;

        // Najpierw sprawdź potencjalną jednostkę
        if (potentialUnit != null && !potentialUnit.isEmpty()) {
            UnitDetectionResult unitInfo = unitService.detectUnitInText(potentialUnit);
            if (unitInfo.isMatch()) {
                unit = unitInfo.getUnit();
                foundKnownUnit = true;
            }
        }

        // Jeśli nie znaleziono jednostki w potencjalnej jednostce,
        // sprawdź w pozostałym tekście
        if (!foundKnownUnit && remainingText != null && !remainingText.isEmpty()) {
            UnitDetectionResult unitInfo = unitService.detectUnitInText(remainingText);
            if (unitInfo.isMatch()) {
                unit = unitInfo.getUnit();
                // Usuwanie jednostki z nazwy produktu
                name = remainingText.replaceAll("(?i)\\b" + Pattern.quote(unit) + "\\b", "").trim();
            }
        }

        // Ustaw jednostkę domyślną, tylko jeśli naprawdę nie znaleziono żadnej innej
        if (unit.isEmpty()) {
            unit = "szt";
        }

        return new UnitProcessingResult(unit, name, true);
    }

    private String cleanProductName(String name) {
        return name
                .replaceAll("\\s*\\([^)]+\\)", "")  // Usuń nawiasy z zawartością
                .replaceAll("\\s+", " ")           // Normalizuj spacje
                .trim()
                .toLowerCase();                    // Konwertuj na małe litery dla spójności
    }

    private boolean isNumeric(String str) {
        if (str == null || str.isEmpty()) {
            return false;
        }
        try {
            Double.parseDouble(str);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }
}